Description: Patches for Debian package creation.
Author: djcj <djcj@gmx.de>
Last-Update: 2014-04-23
--- a/Makefile
+++ b/Makefile
@@ -9,11 +9,3 @@
 check: all
 	test/test.sh
 
-install: all
-	install -m755 embed-ass /usr/bin
-	install -m755 uuencode-simple /usr/bin
-
-install-local: all
-	mkdir -p /usr/local/bin
-	install -m755 embed-ass /usr/local/bin
-	install -m755 uuencode-simple /usr/local/bin
--- a/embed-ass
+++ b/embed-ass
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+#!/bin/bash
 
 # Copyright (c) 2014, djcj <djcj@gmx.de>
 #
@@ -23,19 +23,6 @@
 
 uuencoder="uuencode-simple"
 
-if [ "$(`which $uuencoder`)" = "" ] ; then
-
-    if [ -f $(cd $(dirname "$0") && pwd -P)/"$uuencoder" ] ; then
-
-        uuencoder=$(cd $(dirname "$0") && pwd -P)/"$uuencoder"
-
-    else
-
-        echo "Can't find '$uuencoder'"
-        exit 1
-    fi
-fi
-
 
 if [ "$1" = "" ] || [ "$2" = "" ] ; then
 
